### PR TITLE
Add custom resource grouping

### PR DESCRIPTION
**Pull request by kzlsakal**
Previously, resource name of a span was generated as `<METHOD> <URL>`.

As resource names are often used to categorize resources on APMs, requests
with varying URLs can truly be a pain because they would have different
resource names each time.

This change allows grouping resources with certain patterns such as naming
a request made to `https://website.com/user/1d8f11c8-e2eb-47f2-8f6f-9f58cbdd7c21`
as `https://website.com/user/<user-id>`.

### CODE REVIEW

**Comment by kzlsakal** [comment]
Hi, I had to make these changes for our implementation and wanted to propose this for the upstream in case it can be useful for others.

**Comment by thiamsantos** [comment]
Thanks you very much for the proposal. @kzlsakal 

I think it is really useful we should definitely have some way to configure the resource. 
I just don't like very much the approach of tying it to some custom regex.  

WDYT of having something more flexible without any opinions on how it should be built:

```elixir
SpandexTesla.Telemetry.attach(resource: fn metadata -> metadata[:env][:url] end)
SpandexTesla.Telemetry.attach(resource: &my_fun/1)
``` 

We could just pass the complete metadata returned by tesla, and let the developer choose the best way to name the resource.

Maybe include in the documentation a recommendation to use the middleware `PathParams` to get the template url directly from tesla. https://hexdocs.pm/tesla/Tesla.Middleware.Telemetry.html#module-url-event-scoping-with-tesla-middleware-pathparams-and-tesla-middleware-keeprequest

**Comment by kzlsakal** [comment]
> Thanks you very much for the proposal. @kzlsakal
> 
> I think it is really useful we should definitely have some way to configure the resource.
> I just don't like very much the approach of tying it to some custom regex.
> 
> WDYT of having something more flexible without any opinions on how it should be built:
> 
> ```elixir
> SpandexTesla.Telemetry.attach(resource: fn metadata -> metadata[:env][:url] end)
> SpandexTesla.Telemetry.attach(resource: &my_fun/1)
> ```
> 
> We could just pass the complete metadata returned by tesla, and let the developer choose the best way to name the resource.
> 
> Maybe include in the documentation a recommendation to use the middleware `PathParams` to get the template url directly from tesla. https://hexdocs.pm/tesla/Tesla.Middleware.Telemetry.html#module-url-event-scoping-with-tesla-middleware-pathparams-and-tesla-middleware-keeprequest

I think that's brilliant. I pushed new changes based on your suggestion. Please let me know what you think. Thank you!

File: lib/spandex_tesla.ex [hunk 1]
```diff
@@ -133,8 +142,20 @@ defmodule SpandexTesla do
...
+
+  defp default_resource_name(%{env: %{url: url, method: method}}) do
+    upcased_method = method |> to_string() |> String.upcase()
+    "#{upcased_method} #{url}"
+  end
+
+  defp default_resource_name(%{result: {:ok, %{method: method, url: url}}}) do
+    upcased_method = method |> to_string() |> String.upcase()
+    "#{upcased_method} #{url}"
   end
```
Comment (thiamsantos) [inline lib/spandex_tesla.ex:159]:
```suggestion
  defp default_resource_name(%{env: %{url: url, method: method, opts: opts}}) do
    upcased_method = method |> to_string() |> String.upcase()
    resource_url = Keyword.get(opts, :req_url, url)

    "#{upcased_method} #{resource_url}"
  end

  defp default_resource_name(%{result: {:ok, %{method: method, url: url, opts: opts}}}) do
    upcased_method = method |> to_string() |> String.upcase()
    resource_url = Keyword.get(opts, :req_url, url)

    "#{upcased_method} #{resource_url}"
  end
```

@kzlsakal Maybe for a better behaviour we should support the req_url by default, using it if available. WDYT?

File: README.md [hunk 2]
```diff
@@ -50,6 +49,22 @@ SpandexTesla.Telemetry.attach()
 The docs can
 be found at [https://hexdocs.pm/spandex_tesla](https://hexdocs.pm/spandex_tesla).
 
+## Resource grouping
+
+You can pass a custom resource callback to `SpandexTesla.Telemetry.attach/1` with `:resource` key in the config. If none provided, resource name will default to `<METHOD> <URL>`.
```
Comment (kzlsakal) [inline README.md:54]:
```suggestion
You can pass a custom resource callback to `SpandexTesla.Telemetry.attach/1` with `:resource` key in the config. If none provided, resource name will default to `<METHOD> <REQ_URL>`.
```

File: lib/spandex_tesla.ex (hunk 1 above)
Comment (kzlsakal) [reply lib/spandex_tesla.ex]: That's great! Pushed the changes.

**Review by thiamsantos — APPROVED** [review]

**Comment by thiamsantos** [comment]
Thanks @kzlsakal :metal: :heart:

**Comment by mdepolli** [comment]
This is a great feature, thanks @kzlsakal!

**Pull request merged**